File: src/main/kotlin/at/possnig/aoc25/day8/Day8.kt
package at.possnig.aoc25.day8

import java.io.File
import java.util.TreeSet
import kotlin.math.pow
import kotlin.math.sqrt

fun main(args: Array<String>) {
    val inputFile = "src/main/resources/day8/${args[0]}"
    val numConnections = if (inputFile.endsWith("demo")) 10 else 1000
    File(inputFile).useLines { lines ->
        val points = lines.map { it.split(",")
            .map { str -> str.toInt() }
            .let { ints -> Point(ints[0], ints[1], ints[2]) }
        }.toList()
        val distances = TreeSet<PointPair>()
        for(i in points.indices) {
            for (j in (i + 1)..<points.size) {
                distances.add(PointPair(points[i], points[j]))
            }
        }
        val circuits = mutableMapOf<Point, Int>()
        var connections = 0
        for (distance in distances) {
            val head1 = distance.p.findHead()
            val size1 = circuits[head1]
            val head2 = distance.q.findHead()
            val size2 = circuits[head2]
            if (size1 == null && size2 == null) {
                // new circuit
                distance.q.head = distance.p
                circuits[distance.p] = 2
            } else if (size1 != null && size2 == null) {
                // add 2 to circuit of 1
                distance.q.head = head1
                circuits[head1!!] = size1 + 1
            } else if (size1 == null && size2 != null) {
                // add 1 to circuit of 2
                distance.p.head = head2
                circuits[head2!!] = size2 + 1
            } else if (head1 != head2) {
                // merge circuits
                head2!!.head = head1
                circuits.remove(head2)
                circuits[head1!!] = size1!! + size2!!
            } else {
                // last case: head1 == head2 -> do nothing
            }
            if (++connections == numConnections) {
                circuits.values.sortedByDescending { it }
                    .take(3)
                    .fold(1, Int::times)
                    .also { println("Part 1: $it") }
            }
            if (circuits.size == 1 && circuits.containsValue(points.size)) {
                val result = distance.p.x.toLong() * distance.q.x.toLong()
                println("Part 2: $result")
                break
            }
        }
    }
}

class Point(val x: Int, val y: Int, val z: Int) {

    var head: Point? = null

    fun findHead(): Point? {
        var candidate: Point? = this
        while (candidate?.head != null) {
            candidate = candidate.head
        }
        return candidate
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Point) return false

        if (x != other.x) return false
        if (y != other.y) return false
        if (z != other.z) return false

        return true
    }

    override fun hashCode(): Int {
        var result = x
        result = 31 * result + y
        result = 31 * result + z
        return result
    }
}

data class PointPair(val p: Point, val q: Point): Comparable<PointPair> {
    private var euclid: Double? = null

    fun euclid(): Double {
        synchronized(this) {
            if (euclid == null) {
                euclid =
                    sqrt((p.x.toDouble() - q.x).pow(2) + (p.y.toDouble() - q.y).pow(2) + (p.z.toDouble() - q.z).pow(2))
            }
            return euclid!!
        }
    }

    fun contains(point: Point): Boolean = point == this.p || point == this.q

    override fun compareTo(other: PointPair): Int =
        this.euclid().compareTo(other.euclid())
}

File: src/main/python/day8/day8.py
import math
import operator
import sys
import timeit
from functools import reduce
from typing import Optional


class Point:
    x: int
    y: int
    z: int
    head: 'Point' = None

    def __init__(self, x: int, y: int, z: int) -> None:
        self.x = x
        self.y = y
        self.z = z

    def find_head(self) -> 'Point':
        candidate = self
        while candidate.head is not None:
            candidate = candidate.head
        return candidate

    def __eq__(self, other) -> bool:
        if self is other: return True
        if type(other) is not Point: return False
        return self.x == other.x \
            and self.y == other.y \
            and self.z == other.z

    def __hash__(self):
        result = self.x
        result = 31 * result + self.y
        result = 31 * result + self.z
        return result

    @staticmethod
    def from_string(s: str) -> 'Point':
        ls = list(map(int, s.split(',')))
        return Point(ls[0], ls[1], ls[2])


class PointPair:
    p: Point
    q: Point
    euclid: Optional[float] = None

    def __init__(self, p: Point, q: Point) -> None:
        self.p = p
        self.q = q

    def distance(self):
        if self.euclid is None:
            self.euclid = math.sqrt(
                sum([
                    (self.p.x - self.q.x) ** 2,
                    (self.p.y - self.q.y) ** 2,
                    (self.p.z - self.q.z) ** 2
                ]))
        return self.euclid

    def __contains__(self, item: Point):
        return self.p == item or self.q == item

    def __lt__(self, other):
        return self.distance() < other.distance()


def main() -> None:
    input_file = f'../../resources/day8/{sys.argv[1]}'
    num_connections = 10 if input_file.endswith('demo') else 1000
    with open(input_file, 'r') as h:
        points: list[Point] = list(map(Point.from_string, h.read().split('\n')))
        distances: list[PointPair] = reduce(lambda a, b: a + b, list(map(lambda i: \
                                                                             list(map(lambda j: \
                                                                                          PointPair(points[i],
                                                                                                    points[j]),
                                                                                      range(i + 1, len(points)))),
                                                                         range(len(points)))), [])
        distances.sort()
        circuits: dict[Point, int] = dict()
        connections = 0
        for distance in distances:
            head1 = distance.p.find_head()
            size1 = circuits.get(head1)
            head2 = distance.q.find_head()
            size2 = circuits.get(head2)
            if size1 is None and size2 is None:
                distance.q.head = distance.p
                circuits[distance.p] = 2
            elif size1 is not None and size2 is None:
                distance.q.head = head1
                circuits[head1] = size1 + 1
            elif size1 is None and size2 is not None:
                distance.p.head = head2
                circuits[head2] = size2 + 1
            elif head1 != head2:
                head2.head = head1
                del circuits[head2]
                circuits[head1] = size1 + size2
            else:
                # last case is head1 == head2 -> nothing to do
                pass
            connections += 1
            if connections == num_connections:
                values = list(circuits.values())
                values.sort(reverse=True)
                result = reduce(operator.mul, values[:3])
                print(f'Part 1: {result}')
            if len(circuits) == 1 and len(points) in circuits.values():
                result = distance.p.x * distance.q.x
                print(f'Part 2: {result}')
                break


if __name__ == '__main__':
    execution_time = timeit.timeit(main, number=1)
    print(f'Took {round(execution_time, 4)} Seconds to solve')
